### ISSUE TITLE
runner.http.query constantly causing ERROR 599 - Timeout

### ISSUE THREAD

**Issue by Trouble123** [saltstack/salt#45727]
### Description of Issue/Question
Using runner.http.query sometimes gives HTTP 599 Timeout ERROR. I originally thought it was our web app on Azure, but even when pointing the query to other webhook servers, was getting similar results. The state run is to send 5 grains to a RESTful API.

### Setup
**grains.sls**
```
{% set Items = ["saltversion", "mem_total", "cpuarch", "num_cpus", "osfullname"  ] %}

{% set url = salt.pillar.get('url') %}

{% for item in Items %}
current_{{ item }}:
  grains.present:
    - value: {{ grains[item] }}
    - force: True

update_{{item}}:
  event:
    - send
    - name: app/grains/update
    - data:
        id: {{ grains['id'] }}
        item_name: {{ item }}
        item_value: {{ grains[item] | json }}
        url: {{ url }}

{% endfor %}
```
The reactor file **send_grains.sls**:
```
{% set url = salt.pillar.get('url') %}

webhook1: 
  runner.http.query:
    - arg:
      - {{ data['data']['url'] }}

    - kwarg:
        method: POST
        data:
          id: '{{ data['id'] }}'
          grains: '{ 
                    "{{ data['data']['item_name'] }}": "{{ data['data']['item_value'] }}"}'
```
### Steps to Reproduce Issue
I start salt-master in debug mode. I then run a salt 'MINION' state.apply grains, i see the grains show in the console, then in the salt-master debug screen i see the grains that were able to be sent to the REST API server, such as this:

[DEBUG   ] LazyLoaded nested.output
body:
    ""
[INFO    ] Runner completed: 20180127143329416842
[DEBUG   ] Response Status Code: 200
[DEBUG   ] Sending event: tag = salt/run/20180127143329497579/ret; data = {'fun_args': ['http://<URL>', OrderedDict([('method', 'POST'), ('data', OrderedDict([('id', 'MINION'), ('grains', '{ "osfullname": "Microsoft Windows Server 2012 R2 Datacenter"}')]))])], 'jid': '20180127143329497579', 'return': {'body': '""'}, 'success': True, '_stamp': '2018-01-27T03:33:29.620540', 'user': 'Reactor', 'fun': 'runner.http.query'}
[DEBUG   ] LazyLoaded nested.output

But after 20-25 seconds i normally see 1 or 2 599 errors:

[DEBUG   ] Sending event: tag = salt/run/20180127143329063820/ret; data = {'fun_args': ['http://URL', OrderedDict([('method', 'POST'), ('data', OrderedDict([('id', 'MINION'), ('grains', '{ "saltversion": "2017.7.1"}')]))])], 'jid': '20180127143329063820', 'return': {'status': 599, 'error': 'HTTP 599: Timeout'}, 'success': True, '_stamp': '2018-01-27T03:34:29.126487', 'user': 'Reactor', 'fun': 'runner.http.query'}
[DEBUG   ] LazyLoaded nested.output
error:
    HTTP 599: Timeout
status:
    599

So at this point i wasnt sure if the issue was on the salt end or Azure. I googled the 599 error, and there was some articles about having to change the connect_timeout & request_timeout. I couldnt see the setting anywhere in salt, so after looking around found  the values on line 291/292 of /usr/local/lib/python2.7/dist-packages/tornado/httpclient.py. The default was 20 seconds - so that matched what i was seeing, i changed them to 90 seconds and around 90 seconds after the command was run, i would now see the 599 timeout. Intrestingly i found it was the connect_timeout that mattered, ie if i set the request_timeout to 90, and connect_time at 30, i would get the 599 error around 30 seconds after i saw the runner.http.query being initiated in the salt-master log.

To further validate it being an issue on the salt end, not Azure failing to respond i used Telerik Fiddler.

I setup Telerik Fiddler to pass the data through. So I set it up as a reverse proxy, where all data into Fiddler port 8888 goes out to our web app on Azure.

So i set the URL on salt to point to my PC running telerik fiddler on port 8888, and then setup a fiddler rule to send the data to our REST API on Azure. When i run the state.apply, when i get the 599 error, i see that i only see the corrisponding sessions hit fiddler's UI. So the issue is 100% the salt runner not initiating the http.query for all 5 (or when i used 6 grains 6) queries from the reactor.

Reading the httpclient.py, there is references to using curl_httpclient as its faster, but i am not a developer- so have no idea how to do that. If anyone can point me to how to, will give it a go.

### Versions Report
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: 1.11.2
       cherrypy: unknown
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.3
            ZMQ: 4.0.4
 
System Versions:
           dist: Ubuntu 14.04 trusty
         locale: UTF-8
        machine: x86_64
        release: 3.19.0-25-generic
         system: Linux
        version: Ubuntu 14.04 trusty

**Comment by gtmanfred** [comment]
There have been some updates to the http.query stuff in 2017.7.2, and some more fixes are coming in 2017.7.3, which we are working on getting out right now, would you be able to upgrade to 2017.7.2 and check if that fixes this problem, and if it doesn't can you install from the 2017.7.3 branch?

```
curl -Lo boot https://bootstrap.saltstack.com
bash boot -MN git 2017.7.3
```

The above command will install the 2017.7.3 branch on a new master, and hopefully you would be able to run that and check the `http.query` on a new clean instance.

Thanks,
Daniel

**Comment by Trouble123** [comment]
Hi, i did the upgrade using your 2 line commands and its now showing version: 2017.7.0-n/a-1439da8 
I have tested on this version (can you explain how i can find out what version this is), and the issue does not seem to be happening any more.
Thanks

**Comment by gtmanfred** [comment]
so, when installing from git, _version.py does not get generated, so the version will not actually show correct.  What you installed was directly from the [2017.7.3](https://github.com/saltstack/salt/tree/2017.7.3) branch which we are currently stablizing for release as 2017.7.3 packages.  Since this fixes the issue, I am going to close this issue, and it will be fixed once 2017.7.3 is released.

For more information about the progress of the release, there is a #releases channel on our community slack https://github.com/saltstack/salt#get-saltstack-support-and-help

or you can subscribe to the `salt-announce` mailing list to know when the release actually happens

https://groups.google.com/forum/#!forum/salt-announce

Thanks,
Daniel
